pipelined/bevy_sprite2/src/render/mod.rs

```diff
@@ -219,99 +236,193 @@ struct SpriteVertex {
     pub uv: [f32; 2],
 }
 
+#[repr(C)]
+#[derive(Copy, Clone, Pod, Zeroable)]
+struct ColoredSpriteVertex {
+    pub position: [f32; 3],
+    pub uv: [f32; 2],
+    pub color: [f32; 4],
+}
```

Comment:
If I understand correctly, the current bottleneck is the bandwidth between CPU and GPU. Could we reduce duplicated data by removing position and color from the vertex buffer and instead saving the sprite's transform and color in a separate uniform buffer? (the index in that buffer would then be `vertex_index / 6`)
Memory usage could go from `(3 + 2 + 4) * 6 = 54` bytes per sprite to `2 * 6 + 16 + 4 = 32` bytes per sprite, a 40% reduction.

Reply:
Using a uniform buffer (with a value per-sprite) instead of per-vertex data is ... a clever idea.

I suspect this can be taken even further, by putting everything into the uniform buffer, drawing with no vertex or index buffers, and deriving all the values in the vertex shader. Also, this way, the transform matrix multiply can be done by the GPU.

I suspect this could be the "ultimate" scenario: avoid the low GPU occupancy of instancing, and mitigate the high bandwidth cost of batching.

(Unless I am missing some understanding about why it wouldn't work.)

I am going to play around with this over the next few hours. I want to try implementing it.

If I'm successful, I'll post a separate PR building on top of this one. If you don't hear from me, ignore me and move on. :)

Reply:
Please report back your findings either way. :)

Reply:
OK, so, I haven't written the code yet, but I have some findings to report:

I found that `wgpu`'s implementation of bunnymark implements basically what I suggested above!

[See their vertex shader](https://github.com/gfx-rs/wgpu/blob/master/wgpu-hal/examples/halmark/shader.wgsl)

(though their bunnymark is without batching, they are doing one draw call per sprite, but they are implementing the vertex shader as I described above: with uniform buffers only (for per-sprite data), no vertex or index data, calculating vertex position in the shader)

This gives me hope that this is, indeed, the way to go! It seems optimal. Now, just to adapt this for bevy Bevy, and the batching from this PR...

This should eliminate the bandwidth bottleneck of the batching implementation in this PR and most likely result in much better performance!

Reply:
You're welcome to give this a try. But I also want to keep the merge train moving. I probably won't wait too long to merge this pr, but you're always welcome to follow up with another PR.

Reply:
I actually also started experimenting with vertex-index driven drawing when @StarArawn shared their link on discord, so I might also throw my hat in the ring.

Reply:
Make sure whatever you try still maintains support for texture atlases. The wgpu example just generates default coordinates.

Reply:
One immediate issue that I encountered was that uniform buffers have a max size of 16384 bytes, which isn't enough to drive hundreds of thousands of sprites. This is normally where storage buffers come into play, but WebGL2 doesn't support storage buffers. I think the current impl is the most efficient approach that maintains WebGL2 compatibility, so i think that should be the default. I dont think that should stop us from implemented a faster approach on platforms that support it, but i think we should roll with this impl for now.

Reply:
OK, good to know.

I'd like to throw just one more suggestion to consider: since bandwidth is the bottleneck here ... it may be worth optimizing the data types to represent the per-vertex data in a more compact way?

Some ideas that come to mind:

 - The wgpu shader I linked above stores the sprite color as a single `u32` (i.e integer rather than floating point). For sprites, the color should never be greater than 1.0 anyway, so maybe we can do this, too? It would reduce the size of the color field to 1/4 of its current size. 8-bit precision should be enough...

 - For the UVs, could they be represented as two 16-bit integers (in pixel coordinates) instead of 32-bit floats? The full size of the image/atlas could be placed in a global, and the vertex shader could use it calculate the UVs.
 
 Since we have so many vertices and duplicated data, things like these could result in very significant bandwidth savings.

Reply:
WGSL currently does not support 16 bit integers (which is news to me), so that UV encoding won't work. Using u32 for colors wins us ~2-3k more sprites at 60fps. Its not a huge win, but I do think its worth _slightly_ more internal complexity. Although I do worry a bit about custom sprite vertex shaders, which would need to do the conversion on their own.  However I'm happy to revisit that UX later once custom sprite shader ux has been finalized.

Reply:
Just pushed the u32 color encoding.

Reply:
> WGSL currently does not support 16 bit integers (which is news to me), so that UV encoding won't work.

@cart You could just encode/pack them into a single 32-bit integer, and decode that in the vertex shader with bitmasks and shifts. Similar to the u32 for the color and unpacking that into RGBA.

Though, I guess, maybe too late now, that this PR is merged.

Reply:
Never too late! We can always follow up if it proves to be worth it.